lib/datasource/maven/index.ts

```diff
@@ -136,6 +138,85 @@ function extractVersions(metadata: XmlDocument): string[] {
   return elements.map(el => el.val);
 }
 
+async function getVersionsFromMetadata(
+  dependency: MavenDependency,
+  repoUrl: string
+): Promise<string[] | null> {
+  const metadataUrl = getMavenUrl(dependency, repoUrl, 'maven-metadata.xml');
+  if (!metadataUrl) return null;
+
+  const cacheNamespace = 'datasource-maven-metadata';
+  const cacheKey = metadataUrl.toString();
+  const cachedVersions = await renovateCache.get<string[]>(
+    cacheNamespace,
+    cacheKey
+  );
+  /* istanbul ignore if */
+  if (cachedVersions) return cachedVersions;
+
+  const mavenMetadata = await downloadMavenXml(metadataUrl);
+  if (!mavenMetadata) return null;
+
+  const versions = extractVersions(mavenMetadata);
+  await renovateCache.set<string[]>(cacheNamespace, cacheKey, versions, 10);
+  return versions;
+}
+
+type ArtifactsInfo = Record<string, boolean | null>;
+
+function isValidArtifactsInfo(
+  info: ArtifactsInfo | null,
+  versions: string[]
+): boolean {
+  if (!info) return false;
+  return versions.every(v => info[v] !== undefined);
+}
+
+async function filterMissingArtifacts(
+  dependency: MavenDependency,
+  repoUrl: string,
+  versions: string[]
+): Promise<string[]> {
+  const cacheNamespace = 'datasource-maven-metadata';
+  const cacheKey = dependency.dependencyUrl;
+  let artifactsInfo: ArtifactsInfo | null = await renovateCache.get<
+    ArtifactsInfo
+  >(cacheNamespace, cacheKey);
+
+  if (!isValidArtifactsInfo(artifactsInfo, versions)) {
+    artifactsInfo = {};
+    for (const version of versions) {
```

Comment:
Wondering if doing this in series is too slow

Reply:
Good point, already made it concurrent